clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessor.java

```diff
@@ -206,13 +206,23 @@ public void process(ApplicationEvent event) {
     }
 
     private void process(final PollEvent event) {
+        // Trigger a reconciliation that can safely commit offsets if needed to revoke partitions,
+        // as we're processing before any new fetching starts in the app thread
+        requestManagers.consumerMembershipManager.ifPresent(consumerMembershipManager ->
+            consumerMembershipManager.maybeReconcile(true));
         if (requestManagers.commitRequestManager.isPresent()) {
-            requestManagers.commitRequestManager.ifPresent(m -> m.updateAutoCommitTimer(event.pollTimeMs()));
+            CommitRequestManager commitRequestManager = requestManagers.commitRequestManager.get();
+            commitRequestManager.updateTimerAndMaybeCommit(event.pollTimeMs());
+            // all commit request generation points have been passed,
+            // so it's safe to notify the app thread could start the next poll cycle
+            event.future().complete(null);
             requestManagers.consumerHeartbeatRequestManager.ifPresent(hrm -> {
                 hrm.membershipManager().onConsumerPoll();
                 hrm.resetPollTimer(event.pollTimeMs());
             });
         } else {
+            // unblock if commitRequestManager does not exist
+            event.future().complete(null);
```

Comment:
Without `commitRequestManager`, there is no risk of auto-commit request generation. Additionally, shareConsumer has no auto-commit mechanism, so it's safe to unblock here.